activestorage/test/models/attached/many_test.rb

```diff
@@ -366,6 +366,62 @@ class ActiveStorage::ManyAttachedTest < ActiveSupport::TestCase
     assert_not ActiveStorage::Blob.service.exist?(@user.highlights.second.key)
   end
 
+  test "attaching blobs to an invalid record, keeps attachments in memory" do
+    @user.update_attribute(:name, nil)
+    assert_not @user.valid?
+    assert_not @user.highlights.attached?
+
+    @user.highlights.attach create_blob(filename: "racecar.jpg")
+    assert @user.highlights.attached?
+    assert 1, @user.highlights.count
+    assert_equal "racecar.jpg", @user.highlights.first.filename.to_s
+
+    @user.reload
+    assert_not @user.highlights.attached?
+  end
+
+  test "attaching blobs to an invalid record, can be saved afterwards" do
+    @user.update_attribute(:name, nil)
+    assert_not @user.valid?
+    assert_not @user.highlights.attached?
+
+    @user.highlights.attach create_blob(filename: "racecar.jpg")
+    assert @user.highlights.attached?
+    assert 1, @user.highlights.count
+    assert_equal "racecar.jpg", @user.highlights.first.filename.to_s
+
+    @user.update!(name: "John")
+    @user.reload
+    assert @user.highlights.attached?
+    assert 1, @user.highlights.count
+    assert_equal "racecar.jpg", @user.highlights.first.filename.to_s
+  end
+
+  test "attaching blobs to an invalid record multiple times, keeps some attachments in memory" do
+    @user.update_attribute(:name, nil)
+    assert_not @user.valid?
+    assert_not @user.highlights.attached?
+
+    @user.highlights.attach create_blob(filename: "racecar.jpg")
+    assert @user.highlights.attached?
+    assert 1, @user.highlights.count
+    assert_equal "racecar.jpg", @user.highlights.first.filename.to_s
+
+    @user.highlights.attach create_blob(filename: "funky.jpg"), create_blob(filename: "town.jpg")
+    assert @user.highlights.attached?
+    assert 2, @user.highlights.count
+    assert_equal "racecar.jpg", @user.highlights.first.filename.to_s
+    assert_equal "funky.jpg", @user.highlights.second.filename.to_s
```

Comment:
@jonathanhefner Look at this test 😅 I pushed it so you can see that it's keeping the amount of files but the wrong ones. I suggest removing (or skipping) this test and tackling this bug separately. WDYT?

Reply:
Yes, I would **strongly** prefer to avoid scope creep.  In retrospect, I should have suggested a separate PR in https://github.com/rails/rails/pull/45397#discussion_r901870579.  Would you mind limiting the scope of this PR to the return value of the `attach` methods?

Reply:
There I removed those tests and I thought it'd be valuable to add a CHANGELOG entry, but let me know if you want to adjust the message or even remove it.

Apart from that, I'll work in a separate PR on adding some tests for the "attachments kept in memory" situation and find a fix for the strange scenario I found.

Let me know if I missed something.